lib/SILOptimizer/Transforms/SILCodeMotion.cpp

```diff
@@ -292,6 +293,44 @@ void BBEnumTagDataflowState::handlePredSwitchEnum(SwitchEnumInst *S) {
                    "the switch_enum.");
 }
 
+void BBEnumTagDataflowState::handleStringCmp(SILBasicBlock *Pred) {
+  // Find all apply instructions
+  for (auto begin = Pred->begin(); begin != Pred->end(); ++begin) {
+    if (auto *AI = dyn_cast<ApplyInst>(begin)) {
+      if (auto *FN = dyn_cast<FunctionRefInst>(AI->getCalleeOrigin())) {
+        // Only keep going if this is an apply instruction
+        if (FN->getReferencedFunctionOrNull()->getName() != "$sSS2eeoiySbSS_SStFZ") continue;
+      } else continue;
+      
+      StringRef FirstArg; // Keep track of one of the arguments
+      
+      for (auto& Arg : AI->getArgumentOperands()) {
+        ValueBase *V = (ValueBase*)Arg.get().getOpaqueValue();
+        Operand * Defined = *V->use_begin(); // Find where the argument was defined
+
+        // The string is created using a function that is passed a string literal
+        if (auto *MakeStr = dyn_cast<ApplyInst>(Defined->get())) {
+          // Get the string literal which is the first argument
+          if (auto *SL = dyn_cast<StringLiteralInst>(MakeStr->getArgumentOperands()[0].get())) {
+            if (FirstArg.empty()) {
+              FirstArg = SL->getValue();
+              continue;
+            }
+            
+            APInt IsSame(1, FirstArg == SL->getValue());
```

Comment:
cc @milseman. This is cool! Unfortunately, constant folding string equality is not that cut and dry because it takes Unicode canonical equivalence into account, but this is only going to check that the strings contain the same code units. Without bringing ICU entirely into the compiler itself, we could constant fold cases where the literal is identical as definitely true, and true or false cases where the string is ASCII so we know there are no canonical equivalence issues. Otherwise, we should leave the call as is.

Reply:
Thanks for explaining the issue, that makes sense. I'll add a check to ensure we are looking at an ASCII string literal.

Reply:
This should now be resolved :)

Reply:
> Unfortunately, constant folding string equality is not that cut and dry. Equality takes Unicode canonical equivalence into account, but your check here is only going to check that the strings contain the same code units, so it could incorrectly end up turning things like "e\u{301}" == "é" into false. Without bringing ICU entirely into the compiler itself, we could limit this so that it only constant folds cases where the literal is identical as definitely true, or ASCII literals where we know there are no canonical equivalence issues. Otherwise, we should leave the call as is.

@jckarter @milseman  I have a follow up question about this comparison of unicode strings. I had added support for `string.equals` in the constant evaluator sometime back, which uses `std::string ==` to equate two string literals encountered while evaluating SIL  (which are UTF8 string as I guess we are only using UTF8 for literals in Swift.) It seems `std::string ==` does a byte by byte comparison. Is it correct to do this and fold string literals even in the UTF8 case? It does handle your example `"e\u{301}" == "é"` correctly. It also appears that `_stringCompare` does a  `memcmp` to compare strings that are `fastUTF8`. 

In other words, assuming that all `string_literal ` strings in the  SIL are UTF8, would it suffice to do a raw byte comparison?

Reply:
Yeah, equality tests will have this same problem, so I guess it's a good thing that that the constant evaluator is currently disabled.